topi/python/topi/nn/winograd_util.py

```diff
@@ -0,0 +1,117 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+# pylint: disable=invalid-name
+""" Utility functions for implementing Winograd convolutions
+    * Based on wincnn routines [1] and paper [2]
+      [1] https://github.com/andravin/wincnn
+      [2] Fast Algorithms for Convolutional Neural Networks" Lavin and Gray, CVPR 2016
+"""
+
+from operator import mul
+from functools import reduce
+import numpy as np
+from ..util import const_matrix
+
+
+def _F_m(a, n):
+    f = lambda j, i: reduce(mul, ((a[i]-a[k] if k != i else 1) for k in range(0, n-1)), 1)
+    F = np.fromfunction(np.vectorize(f), (1, n-1), dtype=int)
+    F = np.diagflat(F)
+    F = np.append(F, np.zeros((n-1, 1), dtype=int), axis=1)
+    f = lambda i, j: (1 if j == (n-1) else 0)
+    z = np.fromfunction(np.vectorize(f), (1, n), dtype=int)
+
+    return np.append(F, z, axis=0)
+
+def _A_m(a, m, n):
+    f = lambda i, j: a[i]**j
+    A = np.fromfunction(np.vectorize(f), (m-1, n), dtype=int)
+    f = lambda i, j: (1 if j == (n-1) else 0)
+    z = np.fromfunction(np.vectorize(f), (1, n), dtype=int)
+
+    return np.append(A, z, axis=0)
+
+def _B_m(a, n):
+    f = lambda j, i: reduce(mul, ((a[i]-a[k] if k != i else 1) for k in range(0, n-1)), 1)
+    Ff = np.fromfunction(np.vectorize(f), (1, n-1), dtype=int)
+    f = lambda i, nth: (reduce(mul, [(np.poly1d([1, -a[k]]) if k != i else 1) \
+                        for k in range(0, n-1)], 1)).coef[n-1-nth-1]/Ff[0, i]
+    F = np.fromfunction(np.vectorize(f), (n-1, n-1), dtype=int)
+    f = lambda i, j: -a[i]**(n-1)
+    t = np.fromfunction(np.vectorize(f), (n-1, 1), dtype=int)
+    T = np.append(np.eye(n-1), t, axis=1)
+
+    return np.append(F.T.dot(T), np.array([np.eye(n)[n-1]]), axis=0)
+
+def _cook_toom_convolution(a, n, r):
+    """Compute Cook-Toom convolution A,B,G matrices"""
+
+    alpha = n+r-1
+
+    f = _F_m(a, alpha)
+
+    if f[0, 0] < 0:
+        f[0, :] *= -1
+
+    A = _A_m(a, alpha, n)
+
+    G = _A_m(a, alpha, r).T
+    G = G.dot(np.linalg.inv(f)).T
+
+    B = _B_m(a, alpha)
+    B = B.dot(f.T)
+
+    return (A, B, G)
+
+def _interpolation_points(N):
+    """Propose filter points"""
```

Comment:
Note that the optimal choice of winograd points depends on N.  Mainly, the choice of points for N-1 is not always a subset of points for N.  It also depends on whether 1D or 2D (or 3D) transforms are performed, but I guess in 99% of the cases 2D is used.

Checkout - https://arxiv.org/pdf/1803.10986.pdf - they brute forced search for optimal points.  I have verified their findings.

I've been able to find even better (than the paper above) points, but its still work in progress.

Reply:
@zlateski ,

Thanks for the feedback !
* Well, paper mentioned agrees with actual interpolation points schema from the proposed code:
```
"In general, the points {0, −1, 1, ∞} are good for reducing these costs" ... 
"Problems start to arise where we need more than just these four basic points."
```

* For now, we replace two hardcoded ones with same generated ones, plus we open possibility to more ```tile_sizes```.

@zlateski , I am looking forward for results you obtain, if it is possible let me know, i am willing to join your efforts if it can be materialized into TVM too.